### PR TITLE
Requirements pinning for app code build

### PR DESCRIPTION
**Pull request by emkll**
## Status

Ready for review
## Description of Changes

Towards #3270, #1617 : Pin hashes for securedrop-app-code python requirements. Until now, we were blocked by Trusty tooling (see #3477)

This should be a stopgap until we implement a more complete solution that will also offer .deb reproducibility, described in https://github.com/freedomofpress/securedrop-debian-packaging/issues/34.

## Testing

Test plan:

- [ ] Changes to the app code build logic make sense
- [ ]  All tests pass when running `make build-debs` 
- [ ] .deb build successful and application/infra tests pass (CI should pass)
- [ ] Local upgrade testing is successful
- [ ] Changing hashes for the source tarballs in `securedrop-app-code-requirements.txt` should make the build fail.
- [ ] Modify `securedrop-app-code-requirements.txt` file for a source tarball should result in build failure.

Since the build process is automated, proper verification requires adding a `wait_for: timout=3600s` at the end of `install_files/ansible-base/roles/build-securedrop-app-code-deb-pkg/tasks/build_securedrop_app_code_deb.yml` and using diffoscope to look at build wheels in the wheelhouse vs wheels in the deb package
- [ ] Inside the container, `/tmp/w.out` contains build information of the wheels (confirm that we actually built the wheels)
- [ ] Wheels in `/tmp/securedrop-app-code_0.13.0~rc1+xenial_amd64/var/securedrop/wheelhouse/` are nearly identical to those in the deb package while using diffoscope, the difference is in the zip files and metadata within.

You can unzip the wheels and diff the folder contents, eg:
```
$diffoscope cryptography-2.6.1.dist-info/ ../../cryptography-2.6.1.dist-info/
 |#################################################################|  100%                             Time: 0:00:00 
--- cryptography-2.6.1.dist-info/
+++ ../../cryptography-2.6.1.dist-info/
├── stat {}
│ @@ -1,8 +1,8 @@
│  
│    Size: 4096      	Blocks: 8          IO Block: 4096   directory
│  Links: 2
│  Access: (0755/drwxr-xr-x)  Uid: ( 1000/       m)   Gid: ( 1000/       m)
│  
│ -Modify: 2019-05-15 18:59:21.382879799 +0000
│ +Modify: 2019-05-15 18:57:38.587575536 +0000
│  
[...]
```
and 
```
$ diffoscope cryptography/ ../../cryptography/
 |#################################################################|  100%                             Time: 0:00:03 
--- cryptography/
+++ ../../cryptography/
├── stat {}
│ @@ -1,8 +1,8 @@
│  
│    Size: 4096      	Blocks: 8          IO Block: 4096   directory
│  Links: 4
│  Access: (0755/drwxr-xr-x)  Uid: ( 1000/       m)   Gid: ( 1000/       m)
│  
│ -Modify: 2019-05-15 18:59:21.438879419 +0000
│ +Modify: 2019-05-15 18:57:38.583575562 +0000
│  
│   Birth: -
│   --- cryptography/__about__.py
├── +++ ../../cryptography/__about__.py
├── stat {}
│ │ @@ -1,8 +1,8 @@
│ │  
│ │    Size: 816       	Blocks: 8          IO Block: 4096   regular file
│ │  Links: 1
│ │  Access: (0644/-rw-r--r--)  Uid: ( 1000/       m)   Gid: ( 1000/       m)
│ │  
│ │ -Modify: 2019-02-27 16:02:06.000000000 +0000
│ │ +Modify: 2019-02-28 04:27:52.000000000 +0000
│ │  
│ │   Birth: -
│   --- cryptography/__init__.py
├── +++ ../../cryptography/__init__.py
├── stat {}
│ │ @@ -1,8 +1,8 @@
│ │  
│ │    Size: 527       	Blocks: 8          IO Block: 4096   regular file
│ │  Links: 1
│ │  Access: (0644/-rw-r--r--)  Uid: ( 1000/       m)   Gid: ( 1000/       m)
│ │  
│ │ -Modify: 2019-02-27 16:02:06.000000000 +0000
│ │ +Modify: 2019-02-28 04:27:52.000000000 +0000
│ │  
```

## Deployment

Mostly a dev/build environment change, but all changes to prod instances will be delivered via the securedrop-app-code package.

### If you made non-trivial code changes:

- [x] I have written a test plan and validated it for this PR

### CODE REVIEW

File: securedrop/requirements/securedrop-app-code-requirements.txt [hunk 1]
```diff
@@ -2,43 +2,209 @@
...
+    # via cryptography
+itsdangerous==0.24 \
+    --hash=sha256:cbb3fcf8d3e33df861709ecaf89d9e6629cff0a217bc2848f1b41cd30d360519 \
+    # via flask
+jinja2==2.10.1 \
+    --hash=sha256:065c4f02ebe7f7cf559e49ee5a95fb800a9e4528727aec6f24402a5374c65013 \
+    --hash=sha256:14dd6caf1527abb21f08f86c784eac40853ba93edb79552aa1e4b8aef1b61c7b
+jsmin==2.2.2 \
+    --hash=sha256:b6df99b2cd1c75d9d342e4335b535789b8da9107ec748212706ef7bbe5c2553b
+mako==1.0.7 \
```
Comment (redshiftzero) [inline securedrop/requirements/securedrop-app-code-requirements.txt]: [needs review]

File: securedrop/requirements/securedrop-app-code-requirements.txt [hunk 2]
```diff
@@ -2,43 +2,209 @@
...
-rq==0.10.0
-scrypt==0.8.0
-sh==1.12.14
-six==1.11.0               # via argon2-cffi, cryptography, python-dateutil, qrcode
-sqlalchemy==1.3.3
-typing==3.6.4
-webassets==0.12.1         # via flask-assets
-werkzeug==0.14.1
-wtforms==2.1              # via flask-wtf
+alembic==0.9.9 \
```
Comment (redshiftzero) [inline securedrop/requirements/securedrop-app-code-requirements.txt]: Flagging for the purposes of diff review: there are no securedrop-app-code dependency updates as part of this PR

**Comment by emkll** [comment]
This should be ready for re-review, I've updated the test plan and all tests pass locally.

**Comment by redshiftzero** [comment]
my testing thus far:

- [x] Changes to the app code build logic make sense
- [x] All tests pass when running make build-debs - yes modulo known issue #4191
- [ ] .deb build successful and application/infra tests pass (CI should pass) - waiting on CI
- [ ] Local upgrade testing is successful
- [ ] Changing hashes for the source tarballs in securedrop-app-code-requirements.txt should make the build fail - in progress
- [ ] Inside the container, /tmp/w.out contains build information of the wheels (confirm that we actually built the wheels)
- [ ] Wheels in /tmp/securedrop-app-code_0.13.0~rc1+xenial_amd64/var/securedrop/wheelhouse/ are nearly identical to those in the deb package while using diffoscope, the difference is in the zip files and metadata within.

**Comment by redshiftzero** [comment]
[staging build failure](https://circleci.com/gh/freedomofpress/securedrop/27847) due to too long with no output, gonna restart once..

**Comment by redshiftzero** [comment]
this time [another timeout](https://circleci.com/gh/freedomofpress/securedrop/27857) but later in the provisioning process:

```
    RUNNING HANDLER [common : Wait for server to come back.] ***********************
    ok: [mon-staging -> localhost]
    fatal: [app-staging -> localhost]: FAILED! => {"changed": false, "elapsed": 300, "msg": "Timeout when waiting for search string OpenSSH in 192.168.121.127:22"}
```

**Comment by conorsch** [comment]
- [x] Local upgrade testing is successful 

Performed interactive testing post-upgrade, including:

* submitting documents and messages
* replying to submissions
* viewing replies via source account

LGTM. We're blocked here by recurring CI failures, opened #4440 to track.

**Comment by emkll** [comment]
rebased on latest develop

**Comment by emkll** [comment]
@redshiftzero apologies, this will need another stamp: I had to force push another change to resolve lint failure: the develop-requirements were using bandit 1.6.0, i reverted to 1.4.0  (see https://github.com/freedomofpress/securedrop/issues/4424) .

**Comment by emkll** [comment]
Due to some issues with linting in ci [1,2], I've dropped the commits that would contain the develop-requirements hashes. Let's open a follow up to track.

[1] : https://circleci.com/gh/freedomofpress/securedrop/27948
[2] : https://circleci.com/gh/freedomofpress/securedrop/27927

**Pull request merged**